### PR TITLE
Add giant screenshots

### PR DESCRIPTION
**Pull request by AaronVanGeffen**
This PR introduces the ability to save the entire map as one giant screenshot. Living up to their name, these screenshots are 24,586×12,416 pixels, netting about 30MB each.

### CODE REVIEW

File: src/OpenLoco/Ui/Screenshot.cpp [hunk 1]
```diff
@@ -117,7 +90,117 @@ namespace OpenLoco::Input
...
+                x = centreY + centreX;
+                y = ((centreX - centreY) / 2) - z;
+                break;
+        }
+
+        viewport.viewX = x - ((viewport.viewWidth << zoomLevel) / 2);
+        viewport.viewY = y - ((viewport.viewHeight << zoomLevel) / 2);
+
+        // Ensure sprites appear regardless of rotation
+        // TODO: what is the OpenLoco equivalent?
```
Comment (duncanspumpkin) [inline src/OpenLoco/Ui/Screenshot.cpp:186]: `EntityManager::resetSpatialIndex()`

File: src/OpenLoco/Ui/Screenshot.cpp [hunk 2]
```diff
@@ -117,7 +90,117 @@ namespace OpenLoco::Input
...
+
+        rt.x = 0;
+        rt.y = 0;
+        rt.width = resolutionWidth;
+        rt.height = resolutionHeight;
+        rt.pitch = 0;
+        rt.zoomLevel = 0;
+
+        viewport.render(&rt);
+        free(rt.bits);
```
Comment (duncanspumpkin) [inline src/OpenLoco/Ui/Screenshot.cpp:202]: use after free! You need to prepareSaveScreenshot then free the memory not the other way round!

File: src/OpenLoco/Ui/Screenshot.cpp [hunk 3]
```diff
@@ -117,7 +90,117 @@ namespace OpenLoco::Input
...
+
+        viewport.viewX = x - ((viewport.viewWidth << zoomLevel) / 2);
+        viewport.viewY = y - ((viewport.viewHeight << zoomLevel) / 2);
+
+        // Ensure sprites appear regardless of rotation
+        // TODO: what is the OpenLoco equivalent?
+        // reset_all_sprite_quadrant_placements();
+
+        Gfx::RenderTarget rt{};
+        rt.bits = static_cast<uint8_t*>(malloc(resolutionWidth * resolutionHeight));
```
Comment (duncanspumpkin) [inline src/OpenLoco/Ui/Screenshot.cpp:190]: You can just do a smart pointer here.
```cpp
try
{
    auto data = std::make_unique<uint8_t[]>(resolutionWidth * resolutionHeight);
}
catch (std::bad_alloc&)
{
    return nullptr;
}
rt.bits = data.get();
```
Hmm maybe a bit wordy.

File: src/OpenLoco/Windows/ToolbarTopAlt.cpp [hunk 4]
```diff
@@ -163,18 +164,19 @@ namespace OpenLoco::Ui::Windows::ToolbarTop::Editor
                 break;
 
             case 5:
-            {
-                loco_global<uint8_t, 0x00508F16> _screenshotCountdown;
-                _screenshotCountdown = 10;
+                Input::triggerScreenshotCountdown(10, 0);
+                break;
+
+            case 6:
```
Comment (duncanspumpkin) [inline src/OpenLoco/Windows/ToolbarTopAlt.cpp:170]: Its high time we defined some constants for these.

Comment (AaronVanGeffen) [reply src/OpenLoco/Windows/ToolbarTopAlt.cpp:170]: Perhaps in the same way as the main ToolbarTop window handles these. Good Hacktoberfest issue?

File: src/OpenLoco/Ui/Screenshot.cpp (hunk 2 above)
Comment (AaronVanGeffen) [reply src/OpenLoco/Ui/Screenshot.cpp:203]: Argh! Thanks for spotting that, I'd been looking at this way too long.

File: src/OpenLoco/Ui/Screenshot.cpp [hunk 5]
```diff
@@ -117,6 +91,116 @@ namespace OpenLoco::Input
...
+        viewport.flags = 0;
+
+        const uint16_t centreX = (Map::kMapColumns / 2) * 32 + 16;
+        const uint16_t centreY = (Map::kMapRows / 2) * 32 + 16;
+
+        const uint16_t z = Map::TileManager::getHeight({ centreX, centreY }).landHeight;
+        const auto rotation = main->viewports[0]->getRotation();
+        uint16_t x = 0, y = 0;
+
+        // TODO: I think we have a function for this already?
```
Comment (duncanspumpkin) [inline src/OpenLoco/Ui/Screenshot.cpp:162]: `Map::gameToScreen(pos, rotation)`

File: src/OpenLoco/Ui/Screenshot.cpp [hunk 6]
```diff
@@ -117,6 +91,116 @@ namespace OpenLoco::Input
...
+                x = -centreY + centreX;
+                y = ((-centreX - centreY) / 2) - z;
+                break;
+            case 3:
+                x = centreY + centreX;
+                y = ((centreX - centreY) / 2) - z;
+                break;
+        }
+
+        viewport.viewX = x - ((viewport.viewWidth << zoomLevel) / 2);
```
Comment (duncanspumpkin) [inline src/OpenLoco/Ui/Screenshot.cpp:183]: i dont think the zoomLevel should be here. Its all interms of  the viewport zoom these variables so it should just be viewX = x - viewWidth/ 2

Comment (duncanspumpkin) [reply src/OpenLoco/Ui/Screenshot.cpp:183]: centre2dCoordinates would then make sense for this function. Oh wait i see the issue. Yes do `centre2dCoordinates`.

File: src/OpenLoco/Ui/Screenshot.cpp [hunk 7]
```diff
@@ -117,6 +91,116 @@ namespace OpenLoco::Input
...
+
+        const uint16_t resolutionWidth = ((Map::kMapColumns * 32 * 2) >> zoomLevel) + 8;
+        const uint16_t resolutionHeight = ((Map::kMapRows * 32 * 1) >> zoomLevel) + 128;
+
+        Ui::Viewport viewport{};
+        viewport.width = resolutionWidth;
+        viewport.height = resolutionHeight;
+        viewport.x = 0;
+        viewport.y = 0;
+        viewport.viewWidth = viewport.width;
```
Comment (duncanspumpkin) [inline src/OpenLoco/Ui/Screenshot.cpp:149]: No. This is wrong. It should be scaled by the zoomLevel. `ScreenToViewport::scaleTransform(viewDims, viewport)` would work but need to set the zoom first. Hmm might need to rework `ScreenToViewport` i didn't think it would be useful for stuff that isn't fully constructed.

File: src/OpenLoco/Input.h [hunk 8]
```diff
@@ -102,6 +102,8 @@ namespace OpenLoco::Input
     void setMapSelectionFlags(uint8_t flags);
     void resetMapSelectionFlag(uint8_t flags);
 
+    void triggerScreenshotCountdown(int8_t numTicks, uint8_t type);
```
Comment (duncanspumpkin) [inline src/OpenLoco/Input.h:105]: Create an enum for type.

File: src/OpenLoco/Ui/Screenshot.cpp [hunk 9]
```diff
@@ -117,6 +91,100 @@ namespace OpenLoco::Input
             png_destroy_write_struct(&pngPtr, nullptr);
             throw;
         }
+    }
+
+    // 0x00452667
+    static std::string prepareSaveScreenshot(Gfx::RenderTarget& rt)
```
Comment (AaronVanGeffen) [inline src/OpenLoco/Ui/Screenshot.cpp:97]: Maybe this should just generate an output stream, and leave the writing to the caller. Thoughts?

**Review by duncanspumpkin — APPROVED** [review]
This all looks sensible. I've not tested it but could you add a type for the function to clean it up slightly.
